### PR TITLE
Notification table

### PR DESCRIPTION
**Pull request by derins**
Notifications Page Fixed -

Make sure to uncomment the "use seed_data section" on the bottom of init.py when running flask for testing purposes.

Still needs some way of giving feedback to the user on success calls.

### CODE REVIEW

File: web/web/notifications/static/notifications.js [hunk 1]
```diff
@@ -52,11 +58,26 @@ class Notifications extends React.Component {
...
         refs[id] = e.currentTarget.checked;
         this.setState({checkboxReferences: refs});
-        //this.props.dispatch(updateNotificationPreference(email, notificationType, event.currentTarget.checked));
+        notExist
+        ?
+        this.props.dispatch(action.postNotifications({
+            "alert_type_id" : alertTypeId,
+            "email" : email,
+            "is_active" : refs[id],
+            "created_by" : 1
```
Comment (avpeery) [inline web/web/notifications/static/notifications.js:71]: Not sure if this can be submitted an issue to address later on, but if you do have time to change this I think "created_by" id should correlate with the current logged in user. I believe their id is stored in userData once they log in, you can also get the id with the token on the flask side.

File: web/web/notifications/static/actions.js [hunk 2]
```diff
@@ -39,6 +39,37 @@ export function getNotifications() {
     }
 }
 
+export function postNotifications(data) {
+    return dispatch => {
+        api.post('notification', { json: { ...data } }, (response) => {
```
Comment (avpeery) [inline web/web/notifications/static/actions.js:44]: Not sure where to place this comment but this could also be a future issue, or if you want to address it now. The current logged in user should only see notifications on this page that are created by them. So just filtering the get_notifications so it's only displaying the ones created by the logged in user. (This connects back to the comment about connected the created_by id to the current logged in user). Thanks Derin! Looks great overall.

File: web/web/notifications/static/helpers.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import React from "react";
+import { queue } from "../../static/js/components/app_notification_queue";
+
+export function validateEmail(email) {
```
Comment (avpeery) [inline web/web/notifications/static/helpers.js:4]: I think this could be dried up by using the createPopup function in helpers instead of writing queue.notify again (also capitalization in title is personal preference).

File: web/web/api/v1/notification.py [hunk 4]
```diff
@@ -36,23 +36,23 @@ def get_notifications():
...
-        exclude=['created_by', 'alert_type_id', 'created_at', 'updated_at'])
+        exclude=['created_by', 'created_at', 'updated_at'])
 
     results = notification_schema.dump(notifications, many=True)
 
     return arw.to_json(results)
 
 
-@notifications_api_bp.route('/notification', methods=['PUT'])
+@notifications_api_bp.route('notification',
```
Comment (jongoncalves) [inline web/web/api/v1/notification.py:46]: why did you remove the `/`? Does it actually work w/o it?

File: web/web/models/notification.py [hunk 5]
```diff
@@ -47,6 +48,15 @@ class Notification(Model):
                                        name='_alert_type_id_email_uc'), )
 
     # Methods
+    @validates('email')
+    def empty_string_to_null(self, key, value):
+        # converts empty string to null as field is not nullable
+        if isinstance(value,str) and value == '':
+            return None
+        else:
+            return value
```
Comment (jongoncalves) [inline web/web/models/notification.py:57]: Couple things here that don't make sense to me:
- If the field is not nullable, why are we returning `None`? Is the intention to actually let the DB throw the exception? If so, may be good to comment that
- Why the instance check? Coercion should cover all instance cases
- This check fails if I send a string with multiple whitespaces, like: `'       '` vs `''`, so it may be worth trimming.

File: web/web/package-lock.json [hunk 6]
```diff
@@ -0,0 +1,10074 @@
+{
```
Comment (jongoncalves) [inline web/web/package-lock.json:1]: Where did this package-lock.json come from? You should not be using `npm`. And `yarn` generates a `yarn.lock` and that should be used. Please update your internal dev flow and make sure the readme is accurate.

File: web/web/notifications/static/actions.js (hunk 2 above)
Comment (derins) [reply web/web/notifications/static/actions.js:46]: fixed

File: web/web/notifications/static/notifications.js (hunk 1 above)
Comment (derins) [reply web/web/notifications/static/notifications.js]: fixed

File: web/web/notifications/static/actions.js (hunk 2 above)
Comment (derins) [reply web/web/notifications/static/actions.js:46]: Added a filter in getNotifications() action creator to fix this

File: web/web/api/v1/notification.py (hunk 4 above)
Comment (derins) [reply web/web/api/v1/notification.py]: it works both ways, changed it to make it more consistent with the rest of them

File: web/web/models/notification.py (hunk 5 above)
Comment (derins) [reply web/web/models/notification.py]: fixed

File: web/web/package-lock.json (hunk 6 above)
Comment (derins) [reply web/web/package-lock.json]: removed the file

**Comment by derins** [comment]
Thanks for your comments all!

File: web/web/notifications/static/notifications.js [hunk 7]
```diff
@@ -102,8 +154,38 @@ class Notifications extends React.Component {
...
+                notifications.push({
+                    "alert_type_id" : alertType,
+                    "is_active" : false,
+                    "notification_id" : uuidv4(),
+                    "not_exist" : true
+                })
+            })
+            // sorting items
+            notifications.sort((a, b) => (a.alert_type_id > b.alert_type_id) ? 1 : -1)
+            console.log("okay dude")
```
Comment (avpeery) [inline web/web/notifications/static/notifications.js:186]: :smile:

Comment (derins) [reply web/web/notifications/static/notifications.js:186]: haha thanks :D

**Review by avpeery — APPROVED** [review]
lgtm!

**Review by jongoncalves — APPROVED** [review]
